### PR TITLE
pvr: parse input stream for decoder parameters

### PR DESCRIPTION
**Pull request by FernetMenta**
Having extra data on pvr stream makes Chrystal HD usable with pvr. I think it's a good idea having elupus a look at this.

### CODE REVIEW

File: xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxPVRClient.cpp [hunk 1]
```diff
@@ -103,6 +117,12 @@ void CDVDDemuxPVRClient::Dispose()
     m_streams[i] = NULL;
   }
   m_pInput = NULL;
+
+  if (m_pParser)
+  {
+    m_dllAvCodec.av_parser_close(m_pParser);
+  }
```
Comment (elupus) [inline xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxPVRClient.cpp]: m_pParser must be nulled.

File: xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxPVRClient.cpp [hunk 2]
```diff
@@ -151,6 +231,12 @@ DemuxPacket* CDVDDemuxPVRClient::Read()
     Reset();
   }
 
+  if (m_bNeedExtraData && !ParseVideoPacket(pPacket))
+  {
+    CDVDDemuxUtils::FreeDemuxPacket(pPacket);
+    return CDVDDemuxUtils::AllocateDemuxPacket(0);
+  }
```
Comment (elupus) [inline xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxPVRClient.cpp:272]: Not sure I like that it throws away packets. But maybe the extradata isn't valid for data before that point.

File: xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxPVRClient.cpp [hunk 3]
```diff
@@ -127,6 +147,66 @@ void CDVDDemuxPVRClient::Flush()
...
+          if (i > 0 && i < FF_MAX_EXTRADATA_SIZE)
+          {
+            if (st->ExtraData)
+              delete[] (uint8_t*)(st->ExtraData);
+            st->ExtraSize = i;
+            st->ExtraData = new uint8_t[st->ExtraSize+FF_INPUT_BUFFER_PADDING_SIZE];
+            memcpy(st->ExtraData, pPacket->pData, st->ExtraSize);
+            memset((uint8_t*)st->ExtraData + st->ExtraSize, 0 , FF_INPUT_BUFFER_PADDING_SIZE);
+            m_bNeedExtraData = false;
+            st->type = STREAM_VIDEO;
```
Comment (elupus) [inline xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxPVRClient.cpp]: This is not guaranteed. We should not write this specific for video only. We could need parsing on audio too.

File: xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxPVRClient.cpp [hunk 4]
```diff
@@ -187,6 +273,8 @@ void CDVDDemuxPVRClient::RequestStreams()
       st->iHeight         = props->stream[i].iHeight;
       st->iWidth          = props->stream[i].iWidth;
       st->fAspect         = props->stream[i].fAspect;
+      st->type            = STREAM_NONE;
+      m_bNeedExtraData    = true;
       m_streams[props->stream[i].iStreamIndex] = st;
```
Comment (elupus) [inline xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxPVRClient.cpp]: I'd like parser to be setup here instead.

Comment (elupus) [inline xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxPVRClient.cpp]: or probably better at the end of this function for any marked stream

**Comment by elupus** [comment]
I think the idea is sound. It should be done generic thou, so we can have actual parsers running for all streams not only for split.

**Comment by mikrohard** [comment]
Is there any way to also parse the profile and level?

int iLevel; // encoder level of the stream reported by the decoder. used to qualify hw decoders.
int iProfile; // encoder profile of the stream reported by the decoder. used to qualify hw decoders.

I tried this on the iPad (because videotoolbox also needs extradata). But it still doesn't use the HW decoder because the profile and level info is missing.

The same stream over HTTP is decoded by videotoolbox.

**Comment by FernetMenta** [comment]
@elupus 
Thanks for your comments. What exactly do you mean with: "not only for split."? AFAIK a parser can also be used to concatenate packets. Is this what you mean?
How would i know what kind of streams to parse? Do all steams have extra data?

I was experimenting with av_find_stream_info but this only partly worked: https://github.com/FernetMenta/xbmc/commit/a32321910c59d2d41b3a908962223b04121ef41d
I got lots of ffmpeg errors and finally reduced it just for splitting.

@mikrohard 
 av_find_stream_info would get profile and level but this is quite an expensive operation. Ideally this information would come from the backends as they have to do some parsing anyway.

**Comment by elupus** [comment]
Parser also split streams or invalid split packets into packets that ffmpeg
decoders expect.

It also extract profile, level, bittate and bunch of other things.

**Comment by FernetMenta** [comment]
I would be grateful for a pointer. I was looking into ffmpeg code avformat_find_stream_info and it actually starts decoding in order to get parameters.

**Comment by elupus** [comment]
Skip parameters for now. Just get complete parsing working for any stream
first.

**Comment by FernetMenta** [comment]
I have updated this PR. Audio streams are marked for parsing too, though it has no effect in this iteration because audio streams have to extradata to split.

**Comment by opdenkamp** [comment]
looks sane to me, but this is really something that @elupus should check

**Comment by FernetMenta** [comment]
@opdenkamp 
should I submit this to mainline now?

**Comment by elupus** [comment]
Hold of until pvr is in.

**Comment by FernetMenta** [comment]
Confused. It is already in.

**Comment by opdenkamp** [comment]
@elupus yeah it's already in :)
best send this to mainline yeah

**Comment by elupus** [comment]
Oh :-). Seems i don't get mails on merges of pull requests anymore. Then
yes against master.

**Pull request closed without merging**